draft-irtf-cfrg-hpke.md

```diff
@@ -581,14 +581,20 @@ the left with zero), but MUST return an error if the sequence number
 overflows.
 
 Each encryption or decryption operation increments the sequence
-number for the context in use.  A given context SHOULD be used either
-only for encryption or only for decryption.
+number for the context in use.  A given context MUST be used either
+only for encryption or only for decryption. For example, initiators
```

Comment:
The additions here seem overly restrictive for a low-level document like this.  The underlying hard requirement is that each nonce value MUST be used only once, so in the event of bidirectional comms, the initiator and responder need to coordinate on which nonces belong to which sender.  For example, they might agree that the initiator sends with even-numbered nonces, and the responder with odd-numbered.  Of course, the details of such a scheme are up to the application.  The simplest such scheme is to use a given HPKE interaction for unidirectional communication, so that the initiator's context would only be used for encrypting and the responder's only for decrypting.

Reply:
> they might agree that the initiator sends with even-numbered nonces, and the responder with odd-numbered.

I agree that this is possible, but the document (as specified) doesn't do this. Each sequence number starts at 0 and is always incremented by 1.

Reply:
Ok, I'm coming around on this.  If we're going to lock it down, though, let's lock it down.  Namely, the multi-stage encryption here is just an extension of the single-stage version to allow the sender to stream data:

* Initiator's context MUST be used for encryption only
* Responder's context MUST be used for decryption only
* Any other case, use the exporter

Maybe we should be more explicit that this is streaming PKE (thus unidirectional I->R).  Should we reflect this in the pseudocode?

Reply:
The proposed text works for me! 

> Maybe we should be more explicit that this is streaming PKE (thus unidirectional I->R). Should we reflect this in the pseudocode?

I don't think that's needed. But maybe others will suggest differently?